htdocs/modules/system/blocks/system_blocks.php

```
@@ -689,12 +689,21 @@ function b_system_themes_show($options)
+        // template's onsubmit fallback used by the screenshot-mode submit button.
+        $select->setExtra(' onchange="this.form.elements[\'xoops_theme_redirect\'].value=location.pathname+location.search+location.hash;this.form.submit();" ');
         $themeSelect = $select->render();
     }
 
     $block['theme_select'] = $themeSelect . '<br>(' . sprintf(_MB_SYSTEM_NUMTHEME, '<strong>'
             . count($xoopsConfig['theme_set_allowed']) . '</strong>') . ')<br>';
+    // Seed the return URL from the current request. The block templates also
+    // overwrite this hidden field via an onsubmit handler from window.location,
+    // which keeps the redirect target accurate even when the block is cached
+    // (bcachetime > 0); the server-side seed is the no-JS fallback.
+    $block['theme_redirect'] = \Xmf\Request::getString('REQUEST_URI', '', 'SERVER');
```

PHPMD (2.15.0)

[error] Avoid using static access to class '\Xmf\Request' in method 'b_system_themes_show'. (undefined)

(StaticAccess)